src/us/kbase/test/auth2/service/AuthExternalConfigTest.java

```diff
@@ -0,0 +1,415 @@
+		assertThat("incorrect post link redirect", cfg.getPostLinkRedirect(),
+				is(ConfigItem.noAction()));
+		
+		assertThat("incorrect toMap", cfg.toMap(),
+				is(MapBuilder.<String, ConfigItem<String, Action>>newHashMap()
+						.with("allowedPostLoginRedirectPrefix", ConfigItem.remove())
+						.with("completeLoginRedirect", ConfigItem.set("http://u2.com"))
+						.with("completeLinkRedirect", ConfigItem.set("http://u4.com"))
+						.with("ignoreIPHeaders", ConfigItem.remove())
+						.with("includeStackTraceInResponse", ConfigItem.set("false"))
+						.build()));
+	}
+	
+	@Test
+	public void defaultConfig() {
+		final AuthExternalConfig<Action> cfg = AuthExternalConfig.SET_DEFAULT;
+		
+		assertThat("incorrect trace", cfg.isIncludeStackTraceInResponse(),
+				is(ConfigItem.set(false)));
+		assertThat("incorrect def trace", cfg.isIncludeStackTraceInResponseOrDefault(), is(false));
+		assertThat("incorrect headers", cfg.isIgnoreIPHeaders(), is(ConfigItem.set(false)));
+		assertThat("incorrect def headers", cfg.isIgnoreIPHeadersOrDefault(), is(false));
+		assertThat("incorrect log prefix", cfg.getAllowedLoginRedirectPrefix(),
+				is(ConfigItem.remove()));
+		assertThat("incorrect link redirect", cfg.getCompleteLinkRedirect(),
+				is(ConfigItem.remove()));
+		assertThat("incorrect login redirect", cfg.getCompleteLoginRedirect(),
+				is(ConfigItem.remove()));
+		assertThat("incorrect post link redirect", cfg.getPostLinkRedirect(),
+				is(ConfigItem.remove()));
+		
+		assertThat("incorrect toMap", cfg.toMap(),
+				is(MapBuilder.<String, ConfigItem<String, Action>>newHashMap()
+						.with("allowedPostLoginRedirectPrefix", ConfigItem.remove())
+						.with("completeLoginRedirect", ConfigItem.remove())
+						.with("postLinkRedirect", ConfigItem.remove())
+						.with("completeLinkRedirect", ConfigItem.remove())
+						.with("ignoreIPHeaders", ConfigItem.set("false"))
+						.with("includeStackTraceInResponse", ConfigItem.set("false"))
+						.build()));
+	}
+	
+	@Test
+	public void constructFail() throws Exception {
+		final ConfigItem<URL, Action> setU = ConfigItem.set(new URL("http://f.com"));
+		final ConfigItem<URL, State> staU = ConfigItem.state(new URL("http://f.com"));
+		final ConfigItem<Boolean, Action> setB = ConfigItem.set(true);
+		final ConfigItem<Boolean, State> staB = ConfigItem.state(true);
+		
+		failConstruct(null, setU, setU, setU, setB, setB,
+				new NullPointerException("allowedPostLoginRedirectPrefix"));
+		failConstruct(staU, null, staU, staU, staB, staB,
+				new NullPointerException("completeLoginRedirect"));
+		failConstruct(setU, setU, null, setU, setB, setB,
+				new NullPointerException("postLinkRedirect"));
+		failConstruct(staU, staU, staU, null, staB, staB,
+				new NullPointerException("completeLinkRedirect"));
+		failConstruct(setU, setU, setU, setU, null, setB,
+				new NullPointerException("ignoreIPHeaders"));
+		failConstruct(staU, staU, staU, staU, staB, null,
```

Comment:
Are there generators for this kind of repetitive test code in Java or do you have to do this all by hand?

Reply:
that'd be nice, but no. It's not that bad, just C&P and move the bad input along.